adding new blog (#2)
gAmUssA/kongproxy.io: content/posts/kong-dynamic-plugin-ordering.adoc

---
title: "Kong Dynamic Plugin Ordering"
description: "This is a tutorial how you can use a new dynamic plugin ordering in Kong Gateway."
date: 2022-09-09
draft: false
slug: "plugin-ordering"
authors: ["Viktor Gamov"]
tags: ["Kong", "tutorial"]
categories: ["Kong", "Gateway"]
externalLink : ""
series: []
---
:source-highlighter: rouge
:imagesdir: /images/2022/
:rouge-style: base16
:icons: font
:toc: macro

.TL;DR

> NOTE: Use a new dynamic plugin ordering to control plugins' execution explicitly.

Reviewer: ```suggestion
NOTE: Use https://docs.konghq.com/gateway/latest/[Kong Gateway's] new dynamic plugin ordering to control plugin execution order.
```

You can read more in the official documentation.
TODO add documentation links.

In this blog post/tutorial, you will learn:

- Why plugin ordering is essential.
- How you can use dynamic plugin ordering.

WARNING: This feature is only available in Kong Enterprise starting from version 3.0

<<<

// - 1st paragraph: The problem and recommendation summary. The background can be in an appendix.
// - 2nd paragraph: Further explanation of the situation.
// - 3rd paragraph: The root cause and solution. Other options considered can be in an appendix.
// - 4th paragraph: Full recommendation and implementation plan.

toc::[]

== Introduction of the Problem.

All the bundled Kong Gateway plugins have static execution priority.
It means the `pre-function` plugin, for example, will always be running before `bot-detection` or `zipkin` plugins, `mtls-auth` will always be running before `oauth2`.

> And when customers have a requirement to customize execution order (like run `rate-limiting` before `key-auth`) they can't do that easily.

Reviewer: I'd clearly indicate that this describes the problem pre-3.0 release.

Projects like https://github.com/Kong/priority-updater[`priority-updater`] can wrap existing plugins and update execution order.
But this would be considered a hack and workaround for the actual problem.

In Kong Gateway 3.0 we're introducing a _dynamic plugin ordering_ functionality.
You can specify a directed graph of plugins.
With this, we should be able to specify config that `rate-limiting` should run before `key-auth` or `oath2-introspection`.

== Example: Default behavior before Kong 3.0

.I have Kong Gateway 3.0 Enterprise instance running locally.

> image::kong_gw_3_0_insomnia.png[]

Reviewer: It will be a bit confusing to tell the reader "behavior before Kong 3.0", but then show them an image and text with 3.0

  --data paths=/wow   #<3>

$ curl -i -X POST http://localhost:8001/plugins \
    --data name=key-auth  #<4>

$ curl -i -X POST \
    --url http://localhost:8001/consumers/ \
    --data "username=example-user"  #<5>

$ curl -i -X POST \
  --url http://localhost:8001/consumers/example-user/key-auth/ \
  --data 'key=test'   #<6>

$ curl -i -X POST http://localhost:8001/plugins \
    --data name=rate-limiting \
    --data config.minute=1 \
    --data config.policy=local  #<7>
----
<1> Kong Gateway 3.0 up and running
<2> A service points to a random Wow expression by Owen Wilson.
<3> A route `/wow` points to `wow-service`.
<4> Create a `key-auth` plugin instance.
<5> Create `example-user` consumer.
<6> Create apiKey `test`
<7> Create a global rate limiting plugin.

Here is a default flow: API user calls `:8000/wow` to access API.
Kong (via `key-auth` plugin) is there is an API key in the header.
If there is one, it validates it and passes next to the `rate-limiting` plugin.
If there is no apiKey in the header, Kong throws an error.

[source,json]
.error.json
----
{
    "message": "No API key found in request"
}
----

== A solution implemented in Kong Gateway 3.0

> Next, let's modify this behavior to invoke `rate-limiting` regardless of whether a key is provided or not.

Reviewer: ```suggestion
Next, let's modify this behavior to invoke `rate-limiting` prior to the key-auth plugin.
```
I think it might trip someone up because this might imply that key-auth is only invoked _if_ a key is provided.

[source,bash]
.filename.sh
----
RLID=`curl localhost:8001/plugins | jq -r '.data[] | select(.name == "rate-limiting").id'`

curl -i -X DELETE http://localhost:8001/plugins/$RLID

curl -i -X POST http://localhost:8001/plugins \
  --data name=rate-limiting \
  --data config.minute=1 \
  --data config.policy=local \
  --data ordering.before.access=key-auth    #<1>
----
<1> an essential new config `ordering.before.access` defines what phase of the specified plugin this plugin needs to be invoked.

> Now, when we invoke the `:8000/wow` API, we will get `No API key found in request,` but if we introspect the headers, we will see that the `rate-limiting` plugin is up for the task already.

Reviewer: "inspect" instead of "introspect"

Reviewer: Showing these headers is great. Maybe in the steps above, before you enabled ordering, you could show the result of the same request _without_ the RL headers to drive the point home clearly
